code/modules/cm_aliens/hivebuffs/hivebuff.dm

```diff
@@ -0,0 +1,517 @@
+/datum/hivebuff/proc/_on_cease()
+	_announce_buff_cease()
+	/// Clear refernces to this buff and unregister signal
+	on_cease()
+	if(!ended_via_pylon_qdeletion)
+		for(var/obj/effect/alien/resin/special/pylon/endgame/pylon_to_clear in sustained_pylons)
+			pylon_to_clear.remove_hivebuff()
+			UnregisterSignal(pylon_to_clear, COMSIG_PARENT_QDELETING)
+	LAZYREMOVE(hive.active_hivebuffs, src)
+
+
+/// Checks the number of pylons required and if the hive posesses them
+/datum/hivebuff/proc/_check_num_required_pylons(obj/effect/alien/resin/special/pylon/endgame/purchased_pylon)
+	var/list/viable_pylons = list()
+	if(number_of_required_pylons > 1)
+		for(var/obj/effect/alien/resin/special/pylon/endgame/potential_pylon in hive.active_endgame_pylons)
+			if(potential_pylon == purchased_pylon)
+				continue
+
+			// Pylons can only sustain one buff at a time
+			if(potential_pylon.sustained_buff)
+				continue
+			viable_pylons += potential_pylon
+
+		if(length(viable_pylons) >= (number_of_required_pylons - 1))
+			return TRUE
+		return FALSE
+	return TRUE
+
+/datum/hivebuff/proc/_roundtime_check()
+	if(ROUND_TIME > (SSticker.round_start_time + roundtime_to_enable))
+		return TRUE
+	return FALSE
+
+/// Checks if the hive can afford to purchase the buff returns TRUE if they can purchase and FALSE if not.
+/datum/hivebuff/proc/_check_can_afford_buff()
+	if(hive.buff_points < cost)
+		return FALSE
+
+	return TRUE
+
+/// Checks if this buff is already active in the hive. Returns TRUE if passed FALSE if not.
+/datum/hivebuff/proc/_check_pass_active()
+	for(var/datum/hivebuff/buff as anything in hive.active_hivebuffs)
+		if(src.type == buff.type)
+			return FALSE
+
+	return TRUE
+
+/// Checks if the buff is unique if other buffs are already in use. Return TRUE if passed FALSE if not.
+/datum/hivebuff/proc/_check_pass_unique()
+	if(!is_unique)
+		return TRUE
+
+	if(LAZYLEN(hive.active_hivebuffs))
+		return FALSE
+
+	return TRUE
+
+/// Checks if the buff is reusable and if it's already been used. Returns TRUE if passed, FALSE if not.
+/datum/hivebuff/proc/_check_pass_reusable()
+	if(is_reusable)
+		return TRUE
+
+	for(var/datum/hivebuff/buff as anything in hive.used_hivebuffs)
+		if(type == buff.type)
+			return FALSE
+
+	return TRUE
+
+/// Deducts points from the hive buff points equal to the cost of the buff
+/datum/hivebuff/proc/_purchase_and_deduct(mob/purchasing_mob)
+	if(!_check_can_afford_buff())
+		to_chat(purchasing_mob, SPAN_XENONOTICE("Something went wrong, try again."))
+		return FALSE
+
+	hive.buff_points -= cost
+	return TRUE
+
+/datum/hivebuff/proc/_seek_queen_approval(mob/living/purchasing_mob)
+	if(!hive.living_xeno_queen)
+		return FALSE
+
+	var/mob/living/queen = hive.living_xeno_queen
+	var/queen_response = tgui_alert(queen, "[purchasing_mob] is trying to Purchase [name] at a cost of [cost] [BUFF_POINTS_NAME]. Our hive has [hive.buff_points] [BUFF_POINTS_NAME]. Do you want to approve it?", "Approve Hive Buff", list("Yes", "No"), 20 SECONDS)
```

Comment:
If purchasing mob is queen, we should just omit the information about whos trying to purchase it. Also we ought to also add the `desc` of the boon here so this could be used to figure out what it does in game.